File: backend-python/app/services/pdf_service.py
import os
import tempfile
from typing import Optional
import aiohttp
from PyPDF2 import PdfReader
from opentelemetry import trace

tracer = trace.get_tracer(__name__)

class PDFService:
    def __init__(self):
        self.max_tokens = int(os.getenv("MAX_TOKENS", 1000))
        self.temp_dir = tempfile.gettempdir()

    async def download_and_extract_text(self, doc_id: str, url: str) -> Optional[str]:
        """
        Downloads a PDF file and extracts its text content.
        
        Args:
            doc_id: Document ID to use as filename
            url: URL to download the PDF from
            
        Returns:
            str: Extracted text from the PDF
            
        Raises:
            ValueError: If URL is invalid or file is not a PDF
            IOError: If download or file operations fail
        """
        with tracer.start_as_current_span("download_and_extract_text") as span:
            span.set_attribute("doc_id", doc_id)
            span.set_attribute("url", url)
            
            # Validate URL
            if not url.lower().endswith('.pdf'):
                raise ValueError("URL does not point to a PDF file")
            
            # Create temp file path
            pdf_path = os.path.join(self.temp_dir, f"{doc_id}.pdf")
            
            try:
                # Download file
                async with aiohttp.ClientSession() as session:
                    with tracer.start_span("download_pdf") as download_span:
                        async with session.get(url) as response:
                            if response.status != 200:
                                raise IOError(f"Failed to download PDF: HTTP {response.status}")
                            
                            # Check content type
                            content_type = response.headers.get('content-type', '')
                            if 'application/pdf' not in content_type.lower():
                                raise ValueError(f"Invalid content type: {content_type}")
                            
                            # Save to temp file
                            with open(pdf_path, 'wb') as f:
                                while True:
                                    chunk = await response.content.read(8192)
                                    if not chunk:
                                        break
                                    f.write(chunk)
                
                # Extract text
                with tracer.start_span("extract_text") as extract_span:
                    try:
                        reader = PdfReader(pdf_path)
                        text = ""
                        print(f">>> pages to extract: {len(reader.pages)}")
                        tokens = 0
                        for page in reader.pages:
                            text += page.extract_text() + "\n"
                            # count the words in the text
                            words = text.split()
                            tokens += len(words)
                            if tokens > self.max_tokens:
                                break

                        return text.strip()
                    except Exception as e:
                        raise IOError(f"Failed to extract text from PDF: {str(e)}")
                    
            finally:
                # Clean up temp file
                if os.path.exists(pdf_path):
                    os.remove(pdf_path)

File: backend-dotnet/Services/PdfService.cs
using System.Diagnostics;
using System.Text;
using iText.Kernel.Pdf;
using iText.Kernel.Pdf.Canvas.Parser;
using iText.Kernel.Pdf.Canvas.Parser.Listener;

namespace DocumentQueryService.Api.Services;

/// <summary>
/// Service for PDF downloading and text extraction
/// </summary>
public class PdfService : IPdfService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<PdfService> _logger;
    private readonly int _maxTokens;
    private static readonly ActivitySource ActivitySource = new("DocumentQueryService.PdfService");

    public PdfService(HttpClient httpClient, ILogger<PdfService> logger, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _logger = logger;
        _maxTokens = configuration.GetValue<int>("MAX_TOKENS", 1000);
    }

    public async Task<string?> DownloadAndExtractTextAsync(string docId, string url, CancellationToken cancellationToken = default)
    {
        using var activity = ActivitySource.StartActivity("download_and_extract_text");
        activity?.SetTag("doc_id", docId);
        activity?.SetTag("url", url);

        // Validate URL
        if (!url.ToLowerInvariant().EndsWith(".pdf"))
        {
            throw new ArgumentException("URL does not point to a PDF file", nameof(url));
        }

        var tempPath = Path.GetTempFileName();
        try
        {
            // Download PDF
            using (var downloadActivity = ActivitySource.StartActivity("download_pdf"))
            {
                using var response = await _httpClient.GetAsync(url, cancellationToken);

                if (!response.IsSuccessStatusCode)
                {
                    throw new HttpRequestException($"Failed to download PDF: HTTP {response.StatusCode}");
                }

                // Check content type
                var contentType = response.Content.Headers.ContentType?.MediaType ?? "";
                if (!contentType.Contains("application/pdf", StringComparison.OrdinalIgnoreCase))
                {
                    throw new ArgumentException($"Invalid content type: {contentType}");
                }

                // Save to temp file
                using var fileStream = new FileStream(tempPath, FileMode.Create, FileAccess.Write);
                await response.Content.CopyToAsync(fileStream, cancellationToken);
            }

            // Extract text
            using var extractActivity = ActivitySource.StartActivity("extract_text");
            try
            {
                using var pdfReader = new PdfReader(tempPath);
                using var pdfDoc = new PdfDocument(pdfReader);

                var text = new StringBuilder();
                var tokenCount = 0;

                for (int i = 1; i <= pdfDoc.GetNumberOfPages(); i++)
                {
                    var page = pdfDoc.GetPage(i);
                    var strategy = new SimpleTextExtractionStrategy();
                    var pageText = PdfTextExtractor.GetTextFromPage(page, strategy);

                    text.AppendLine(pageText);

                    // Count words (rough token approximation)
                    var words = pageText.Split(new[] { ' ', '\t', '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
                    tokenCount += words.Length;

                    if (tokenCount > _maxTokens)
                    {
                        _logger.LogInformation("Reached maximum token limit of {MaxTokens} for document {DocId}", _maxTokens, docId);
                        break;
                    }
                }

                var extractedText = text.ToString().Trim();

                extractActivity?.SetTag("extracted_length", extractedText.Length);
                extractActivity?.SetTag("token_count", tokenCount);

                return extractedText;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to extract text from PDF {DocId}", docId);
                throw new InvalidOperationException($"Failed to extract text from PDF: {ex.Message}", ex);
            }
        }
        finally
        {
            // Clean up temp file
            try
            {
                if (File.Exists(tempPath))
                {
                    File.Delete(tempPath);
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to delete temporary file {TempPath}", tempPath);
            }
        }
    }
}